File: matlab/src/api/mc/ctmc_hitting_time.m
function h = ctmc_hitting_time(Q, targetStates)
% H = CTMC_HITTING_TIME(Q, TARGETSTATES)
%
% Mean time to reach any state in TARGETSTATES from each state of a CTMC with
% generator Q. Target states have zero hitting time; a state that cannot reach
% the set has an infinite one.
%
% Continuous-time twin of DTMC_HITTING_TIME, and the first-moment special case
% of CTMC_PASSAGE_MOMENTS: (-S) h = 1 on the non-target block, where
% DTMC_HITTING_TIME solves (I - P_NT) h = 1. TARGETSTATES is 1-based here and
% 0-based in the Java, Python and C++ twins.
%
% Copyright (c) 2012-2026, Imperial College London
% All rights reserved.

% MALL does not depend on the initial law, so a uniform one is passed rather
% than paying for the stationary solve that PI0 = [] would trigger.
n = size(Q,1);
mall = ctmc_passage_moments(Q, ones(1,n)/n, targetStates, 1);
h = mall(:,1);
end

File: jar/src/main/java/jline/api/mc/Ctmc_hitting_time.java
/*
 * Copyright (c) 2012-2026, QORE Lab, Imperial College London
 * All rights reserved.
 */
package jline.api.mc;

import jline.util.matrix.Matrix;

/**
 * Mean time to reach any state in a target set from each state of a CTMC.
 *
 * <p>Continuous-time twin of {@link Dtmc_hitting_time} and the first-moment
 * special case of {@link Ctmc_passage_moments}: (-S) h = 1 on the non-target
 * block, where {@code dtmc_hitting_time} solves (I - P_NT) h = 1. Target states
 * have zero hitting time; a state that cannot reach the set has an infinite one.
 *
 * <p>Target indices are 0-based here and 1-based in the MATLAB reference.
 */
public final class Ctmc_hitting_time {

    private Ctmc_hitting_time() {
    }

    public static Matrix ctmc_hitting_time(Matrix Q, int[] targetStates) {
        int n = Q.getNumRows();
        // mall does not depend on the initial law, so a uniform one is passed
        // rather than requiring the caller to invent one.
        Matrix pi0 = new Matrix(1, n);
        for (int i = 0; i < n; i++) {
            pi0.set(0, i, 1.0 / n);
        }
        PassageMomentsResult pm =
                Ctmc_passage_moments.ctmc_passage_moments(Q, pi0, targetStates, 1);
        Matrix h = new Matrix(n, 1);
        for (int i = 0; i < n; i++) {
            h.set(i, 0, pm.mall.get(i, 0));
        }
        return h;
    }
}
